CodingChallenge/MusicSearch/MusicSearchViewController.swift

```diff
@@ -0,0 +1,136 @@
+//  you can apply clean architecture to your iOS and Mac projects,
+//  see http://clean-swift.com
+//
+
+import UIKit
+
+
+// Enum to display seach status
+private enum SearchDisplayResults: String {
+    case firstMessage = "Enter Search Term"
+    case loadingMessage = "Searching"
+    case noResults = "No Results Found"
+}
+
+protocol MusicSearchDisplayLogic: class {
+    func displayResults(viewModel: MusicSearch.Search.ViewModel)
+}
+
+class MusicSearchViewController: UIViewController, MusicSearchDisplayLogic, UITextFieldDelegate, UICollectionViewDelegate {
+    
+    @IBOutlet weak var searchTextField: UITextField!
+    @IBOutlet weak var cancelButton: UIButton!
+    @IBOutlet weak var searchStatusLabel: UILabel!
+    @IBOutlet weak var resultsCollection: UICollectionView!
+    
+    let resultsDataSource = SearchDataSource()
+    
+    var interactor: MusicSearchBusinessLogic?
+    var router: (NSObjectProtocol & MusicSearchRoutingLogic & MusicSearchDataPassing)?
+    
+    // MARK: Object lifecycle
+    
+    override init(nibName nibNameOrNil: String?, bundle nibBundleOrNil: Bundle?) {
+        super.init(nibName: nibNameOrNil, bundle: nibBundleOrNil)
+        setup()
+    }
+    
+    required init?(coder aDecoder: NSCoder) {
+        super.init(coder: aDecoder)
+        setup()
+    }
+    
+    // MARK: Setup
+    
+    private func setup() {
+        let viewController = self
+        let interactor = MusicSearchInteractor()
+        let presenter = MusicSearchPresenter()
+        let router = MusicSearchRouter()
+        viewController.interactor = interactor
+        viewController.router = router
+        interactor.presenter = presenter
+        presenter.viewController = viewController
+        router.viewController = viewController
+        router.dataStore = interactor
+    }
+    
+    // MARK: Routing
+    
+    override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
+        if segue.identifier == "ShowSong" {
+            let data = sender as? SongResponse
+            let vc = segue.destination as! SongDetailViewController
+            vc.albumURL = data?.artworkUrl100 ?? ""
+            vc.artist = data?.artistName ?? ""
+            vc.track = data?.trackName ?? ""
+        }
+    }
+    
+    // MARK: View lifecycle
+    
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        
+        resultsCollection?.dataSource = resultsDataSource
+        resultsCollection?.delegate = self
+        resultsCollection?.register(UINib.init(nibName: "SearchResultCell", bundle: nil), forCellWithReuseIdentifier: "SearchResultCell")
+        resultsCollection?.isHidden = true
+        cancelButton?.isEnabled = false
+        searchStatusLabel?.text = SearchDisplayResults.firstMessage.rawValue
+    }
+    
+    // MARK: Search Function
+    func performSearch() {
+        let request = MusicSearch.Search.Request(searchTerm: searchTextField?.text ?? "")
```

Comment:
If the searchTextField is blank, you can probably just skip performing the search or provide default behavior to the search that clears search results.